enhancements/olm-crd-conversion-webhooks.md

```diff
@@ -0,0 +1,94 @@
+status: implementable
+---
+
+# OLM CRD Conversion Webhook Support
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement covers the necessary steps to add OLM support for [conversion](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#webhook-conversion) webhooks.
+
+## Definitions
+
+- Webhook: Webhooks are HTTP callbacks, providing a way for notifications to be delivered to an external web server i.e. a simple event-notification via HTTP POST. Examples: admission webhooks, CRD conversion webhooks
+
+- Conversion Webhook: Webhook that can convert an object from one version to another. 
+
+  When API clients, like kubectl or your controller, request a particular version of your resource, the Kubernetes API server needs to return a result that’s of that version. However, that version might not match the version stored by the API server.
+
+  In that case, the API server needs to know how to convert between the desired version and the stored version. Since the conversions aren’t built in for CRDs, the Kubernetes API server calls out to a webhook to do the conversion instead. 
+
+## Motivation
+
+Many Operators are shipping conversion webhooks to deliver conversion of an object from one version to another. The CustomResourceDefinition API supports a versions field that is used to support multiple versions of custom resources that have been developed. Versions can have different schemas and a conversion webhook is used to convert custom resources between versions.
+
+OLM needs to support operators with crd conversion webhooks and help manage and rotate corresponding certificates.
+
+### Goals
+
+- Operator authors can include crd conversion webhooks in their bundle as a part of a CSV.
+- Operator author will not need to manage [cert](https://www.ssl.com/faqs/what-is-a-certificate-authority/) rotation for webhooks shipped with their operators.
+
+## Proposal
+
+Operator authors will be able to:
+
+- Provide a CRD name in the WebhookDescription, and this CRD would have the conversion webhook defined. OLM would update the CRD with correct CA cert.
+
+
+### CSV support for CRD Conversion Webhook
+
+- The WebhookDescription struct would include one extra field called ‘ConversionCrd’ in the WebhookDescription struct which would hold the CRD name. 
+
+```
+type WebhookDescription struct {
+    .
+    .
+    .
+    // ConversionCrd must match the CustomResourceDefinition name
+    ConversionCrd           string              `json:"conversionCrd"`
+}
+```
+
+- Cluster admins will be allowed to install CRD conversion webhooks for only AllNamespace Operators.
```

Comment:
can it be for operators that have both single namespace and all namespaces? or exclusively all namespace? If it is a restriction, please put it in the non-goal section right under the goal. ie: non-goal: allow CRD conversion webhooks for all install modes. 

This way, people look at it and realize the restrictions right away instead of looking for it in the enhancement.

Reply:
This enhancement is for AllNamespace Operators only. I'll add this under non-goals. Thanks!